hbase-server/src/main/java/org/apache/hadoop/hbase/procedure2/store/region/RegionProcedureStore.java

```diff
@@ -0,0 +1,391 @@
+          procs.add(proto);
+          maxProcId = Math.max(maxProcId, proto.getProcId());
+        }
+      } while (moreRows);
+    }
+    loader.setMaxProcId(maxProcId);
+    ProcedureTree tree = ProcedureTree.build(procs);
+    loader.load(tree.getValidProcs());
+    loader.handleCorrupted(tree.getCorruptedProcs());
+  }
+
+  private void serializePut(Procedure<?> proc, List<Mutation> mutations, List<byte[]> rowsToLock)
+    throws IOException {
+    ProcedureProtos.Procedure proto = ProcedureUtil.convertToProtoProcedure(proc);
+    byte[] row = Bytes.toBytes(proc.getProcId());
+    mutations.add(new Put(row).addColumn(FAMILY, PROC_QUALIFIER, proto.toByteArray()));
+    rowsToLock.add(row);
+  }
+
+  private void serializeDelete(long procId, List<Mutation> mutations, List<byte[]> rowsToLock) {
+    byte[] row = Bytes.toBytes(procId);
+    mutations.add(new Delete(row));
+    rowsToLock.add(row);
+  }
+
+  @Override
+  public void insert(Procedure<?> proc, Procedure<?>[] subProcs) {
+    if (subProcs == null || subProcs.length == 0) {
+      // same with update, just insert a single procedure
+      update(proc);
+      return;
+    }
+    List<Mutation> mutations = new ArrayList<>(subProcs.length + 1);
+    List<byte[]> rowsToLock = new ArrayList<>(subProcs.length + 1);
+    try {
+      serializePut(proc, mutations, rowsToLock);
+      for (Procedure<?> subProc : subProcs) {
+        serializePut(subProc, mutations, rowsToLock);
+      }
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to insert proc {}, sub procs {}", proc,
+        Arrays.toString(subProcs), e);
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void insert(Procedure<?>[] procs) {
+    List<Mutation> mutations = new ArrayList<>(procs.length);
+    List<byte[]> rowsToLock = new ArrayList<>(procs.length);
+    try {
+      for (Procedure<?> proc : procs) {
+        serializePut(proc, mutations, rowsToLock);
+      }
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to insert procs {}", Arrays.toString(procs), e);
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void update(Procedure<?> proc) {
+    try {
+      ProcedureProtos.Procedure proto = ProcedureUtil.convertToProtoProcedure(proc);
+      region.put(new Put(Bytes.toBytes(proc.getProcId())).addColumn(FAMILY, PROC_QUALIFIER,
+        proto.toByteArray()));
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to update proc {}", proc, e);
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void delete(long procId) {
+    try {
+      region.delete(new Delete(Bytes.toBytes(procId)));
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to delete proc id={}", procId, e);
```

Comment:
It is 'pid', not 'proc id'. When we label items differently, it makes it hard tracing history of Procedures.